### PR TITLE
Fix courses management & omniauth tests

### PR DESCRIPTION
**Pull request by allenwq**
`merge(CourseUser.owner).eager_load(:instance, course_users: :user)`  

The SQL generated by this line filters out the courses which do not have an owner. Which leads to tests failing.

### CODE REVIEW

**Comment by lowjoel** [comment]
What does merge(CourseUser.owner) actually do? owner is a field right?

**Comment by allenwq** [comment]
Yes, but it's also a scope.

**Comment by allenwq** [comment]
@lowjoel omniauth tests also fixed the

File: spec/controllers/user/omniauth_callbacks_controller_spec.rb [hunk 1]
```diff
@@ -28,6 +28,7 @@
       context 'when the data provided is incomplete' do
         before do
           facebook_data.info.email = nil
+          facebook_data.uid = SecureRandom.hex
```
Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: ...why random?

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: So that the user created by this uid in other tests won't affects this test.

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: then you should fix the mock data in the let block at line 10?

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: The first 2 tests want the same UID, maybe I should scope them in one context ?

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: if they need the same UID they should share some variable.

**Comment by allenwq** [comment]
@lowjoel How about it now ?

**Comment by lowjoel** [comment]
Actually I saw a mock response in spec/support. Maybe that should be changed instead?

I don't know the design of omniauth nor its specs setup.

**Comment by allenwq** [comment]
@lowjoel That one is the default value (if you don't config request.env in your tests), and only generated once.

File: spec/controllers/user/omniauth_callbacks_controller_spec.rb [hunk 2]
```diff
@@ -5,24 +5,34 @@
...
       before { controller.request.env['omniauth.auth'] = facebook_data }
       subject { get :facebook }
 
-      it 'creates a user identity' do
-        subject
-        identity = User::Identity.find_by(provider: facebook_data.provider, uid: facebook_data.uid)
-        expect(identity).to be_present
-      end
+      context 'when required data is provided' do
+        let(:uid) { self.class::UID }
```
Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: then how is this different from the mock data?

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: Oh, the data configured is only available in the feature tests. you should do this in controller test.

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: but you're accessing the same data =.=

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: Eh, why not ?  Otherwise I need to manually create it `OmniAuth::AuthHash.new()`

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: You said it's for feature specs, but you're accessing that data from a controller spec, so what you're saying doesn't make sense to me...?

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: In controller test we need to get the data and manually assign it to `request.env['omniauth.auth']`, most of the data is reusable.

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: I'm not referring to you needing to assign the value to request.env. I'm saying you are using mock data in your spec that you said was for feature specs.

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: \>< I think I'm not precise enough. It's just that the data will be set automatically in future tests, and future tests need it. In controller tests, you can either use the configured data or manually create your own new data.

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: Maybe we should change this into a factory?

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: This assignment and dup'ing looks very confusing to me.

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: Sure !

File: spec/support/omniauth.rb [hunk 3]
```diff
@@ -2,15 +2,10 @@
...
-          uid: SecureRandom.random_number(2**48),
-          info: {
-            name: FactoryGirl.generate(:name),
-            email: FactoryGirl.generate(:email)
-          }
-        )
     end
   end
+
+  config.before(:each) do
```
Comment (lowjoel) [inline spec/support/omniauth.rb]: Any way to restrict this to only for feature specs? calling like this for every model spec is wasteful

File: spec/support/omniauth.rb [hunk 4]
```diff
@@ -2,15 +2,10 @@
...
-          uid: SecureRandom.random_number(2**48),
-          info: {
-            name: FactoryGirl.generate(:name),
-            email: FactoryGirl.generate(:email)
-          }
-        )
     end
   end
+
+  config.before(:each, type: :feature) do
```
Comment (allenwq) [inline spec/support/omniauth.rb]: @lowjoel Good idea, I've updated !

File: spec/controllers/user/omniauth_callbacks_controller_spec.rb [hunk 5]
```diff
@@ -5,24 +5,34 @@
 
   with_tenant(:instance) do
     before { controller.request.env['devise.mapping'] = Devise.mappings[:user] }
+    self::UID = SecureRandom.random_number(2**48)
 
     describe '#facebook' do
-      let(:facebook_data) { OmniAuth.config.mock_auth[:facebook].dup }
+      let(:uid) { SecureRandom.random_number(2**48) }
+      let(:facebook_data) do
+        data = generate(:omniauth_facebook)
```
Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: generate(..., uid: uid)

Comment (allenwq) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: I think sequence generator does not support arguments ?

Comment (lowjoel) [inline spec/controllers/user/omniauth_callbacks_controller_spec.rb]: then replace the sequence with a normal factory. factories do not need to generate ActiveRecord objects, the AuthHash should be able to be created in a factory.

**Comment by allenwq** [comment]
@lowjoel Updated, I thought FactoryGirl.lint would fail if I do that, but actually it doesn't

File: spec/factories/omniauth_providers.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+FactoryGirl.define do
+  factory :omniauth_provider, class: OmniAuth::AuthHash.name do
+    provider 'default'
+    uid SecureRandom.random_number(2**48)
+    info do
+      {
+        name: FactoryGirl.generate(:name),
```
Comment (lowjoel) [inline spec/factories/omniauth_providers.rb]: Do we need the FactoryGirl prefix? generate(:name) should work right?

**Comment by allenwq** [comment]
@lowjoel It's needed in `before(:suite)`, so that's why it was there previously. I just removed it

**Comment by lowjoel** [comment]
@allenwq you didn't push!

**Pull request merged**